src/modules/Game.class.js

```diff
@@ -0,0 +1,254 @@
+          emptyIndexes.unshift(cellIndex);
+          this.#score += row[previousCellIndex];
+          isStateChanged = true;
+
+          if (row[previousCellIndex] >= 2048) {
+            this.#status = 'win';
+          }
+        }
+      }
+    });
+
+    if (isStateChanged) {
+      this.#addNumber();
+    }
+
+    if (!this.#isForCheck && !this.#isMovePossible()) {
+      this.#status = 'lose';
+    }
+
+    return isStateChanged;
+  }
+  moveRight() {
+    this.#reverse();
+
+    const result = this.moveLeft();
+
+    this.#reverse();
+
+    return result;
+  }
+  moveUp() {
+    this.#rotate();
+
+    const result = this.moveLeft();
+
+    this.#rotate();
+
+    return result;
+  }
+  moveDown() {
+    this.#rotate();
+
+    const result = this.moveRight();
+
+    this.#rotate();
+
+    return result;
+  }
+
+  /**
+   * @returns {number}
+   */
+
+  getScore() {
+    return this.#score;
+  }
+
+  /**
+   * @returns {number[][]}
+   */
+
+  getState() {
+    return this.#state.map((row) => [...row]);
+  }
+
+  /**
+   * Returns the current game status.
+   *
+   * @returns {string} One of: 'idle', 'playing', 'win', 'lose'
+   *
+   * `idle` - the game has not started yet (the initial state);
+   * `playing` - the game is in progress;
+   * `win` - the game is won;
+   * `lose` - the game is lost
+   */
+  getStatus() {
+    return this.#status;
+  }
+
+  /**
+   * Starts the game.
+   */
+  start() {
+    this.#addNumber();
+    this.#addNumber();
+    this.#status = 'playing';
+  }
+
+  /**
+   * Resets the game.
+   */
+  restart() {
+    this.#applyInitialState();
+    this.#score = 0;
+    this.#status = 'idle';
+  }
+
+  #applyInitialState() {
+    this.#state = this.#initialState.map((row) => [...row]);
+  }
+  /**
+   * Checks if any move is possible
+   * @returns {boolean}
+   */
+  #isMovePossible() {
+    const gameCopy = new Game(this.getState(), true);
+
+    return (
+      gameCopy.moveLeft() ||
+      gameCopy.moveRight() ||
+      gameCopy.moveUp() ||
+      gameCopy.moveDown()
+    );
+  }
+  /**
+   * Returns if adding number to state was succesfull
+   * @returns {boolean}
+   */
+  #addNumber() {
+    const number = Math.random() > 0.1 ? 2 : 4;
+    const emptyFields = this.#findEmptyFields();
+
+    if (!emptyFields.length) {
+      return false;
+    }
+
+    const randomIndex = Math.floor(Math.random() * emptyFields.length);
+    const { x, y } = emptyFields[randomIndex];
+
+    this.#state[y][x] = number;
+
+    return true;
+  }
+  /**
+   * Returns positions of empty fields in state
+   * @returns {({ x: number, y: number})[]}
+   */
+  #findEmptyFields() {
+    const result = [];
+
+    this.#state.forEach((row, y) => {
+      row.forEach((cell, x) => {
+        if (cell === 0) {
+          result.push({ x, y });
+        }
+      });
+    });
+
+    return result;
+  }
+
+  #reverse() {
+    this.#state.forEach((row) => row.reverse());
+  }
+
+  #rotate() {
+    this.#state = this.#state.map((row, i) => {
```

Comment:
The '#rotate' method assumes a square matrix. If the game board can be non-square, additional validation is needed to ensure that the rotation is performed correctly.